Manager/ORMManager.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+namespace WebtownPhp\BannerBundle\Manager;
+
+use Symfony\Bridge\Doctrine\RegistryInterface;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use WebtownPhp\BannerBundle\Entity\Banner;
+use WebtownPhp\BannerBundle\Entity\BannerRepository;
+use WebtownPhp\BannerBundle\Event\BannerEmptyEvent;
+use WebtownPhp\BannerBundle\Event\BannerSelectEvent;
+use WebtownPhp\BannerBundle\Event\Events;
+
+class ORMManager implements ManagerInterface
+{
+    /**
+     * @var RegistryInterface
+     */
+    protected $doctrine;
+
+    /**
+     * @var EventDispatcher
+     */
+    protected $dispatcher;
+
+    /**
+     * ORMManager constructor.
+     *
+     * @param RegistryInterface $doctrine
+     */
+    public function __construct(RegistryInterface $doctrine, EventDispatcher $dispatcher)
+    {
+        $this->doctrine = $doctrine;
+        $this->dispatcher = $dispatcher;
+    }
+
+    /**
+     * Get banner to given place
+     *
+     * @param string    $placeName  Banner place name
+     * @param bool|true $isMeasured Should views be counted
+     */
+    public function getBannerTo($placeName, $isMeasured = true)
+    {
+        if ($banner = $this->doGetBanner($placeName)) {
+            $event = new BannerSelectEvent();
+            $name = Events::BANNER_SELECT;
+        } else {
+            $event = new BannerEmptyEvent();
+            $name = Events::BANNER_EMPTY;
+        }
+        $this->dispatcher->dispatch($name, $event);
+
+        if ($event instanceof BannerSelectEvent) {
+          if ($event->getIsMeasured()) {
+              $this->measureShowCount($banner);
+          }
+        }
+    }
+
+    /**
+     * @param string $placeName
+     *
+     * @return Banner
+     */
+    protected function doGetBanner($placeName)
+    {
+        $banners = $this->getRepository()->getBannersForPlace($placeName);
+        $sum = $this->getRepository()->getPrioritySumForPlace($placeName);
+
+        return $this->doGetByPriority($banners, $sum);
+    }
+
+    /**
+     * Get a random banner by priority
+     *
+     * @param array $banners
+     * @param int   $max
+     *
+     * @return Banner
+     */
+    protected function doGetByPriority(array $banners, $max)
```

Comment:
Ez így gyakorlatilag jó, elméletileg nem :) Minden függvénynek illik a saját hatáskörében egészet alkotni. A `$max` az valójában itt nem egy változó, nem fordulhat olyan elő, hogy ugyanannál a `$banners` értéknél eltérő `$max`-okat használunk, a `$banners` ráadásul tartalmazza a szükséges adatokat, ez felesleges "duplikálás". Elég paraméterként a `$banners`, majd egy `foreach()`-csel be kell járni, ki kell számolni itt függvényen belül a max-ot, és után a végén nem `null`-t kell visszaadni, hanem exception-t kell dobni, hiszen elméletileg olyan történt, ami nem történhet meg --> értesítést kell róla kapnunk, hogy az algoritmusban hiba van, rosszul működik.

És ha ez megvan, akkor írni kell rá tesztet. A `rand()` miatt nehéz az ilyenekre tesztet írni, de nem lehetetlen. A lényeg, hogy egy `for()` ciklusban ellenőrizni kell, hogy pl 0-1 priority értéknél, 100 hívás esetén nem kapunk választ a 0-s esetén, és 1-1 esetén az arányuk `0,4` és `0,6` között alakul pl. Vagy igazából ez egyébként is gyors, lehet 100 helyett 10000 is a ciklus esetleg.